barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/generated/AvmMini_trace.cpp

```diff
@@ -0,0 +1,346 @@
+// TODO: taking care of intermediate register values consistency and propagating their
+// values to the next row when not overwritten.
+void AvmMiniTraceBuilder::callDataCopy(uint32_t s0, uint32_t s1, uint32_t d0, std::vector<FF> const& callDataMem)
+{
+    // We parallelize storing memory operations in chunk of 3, i.e., 1 per intermediate register.
+    // This offset points to the first storing operation (pertaining to intermediate register Ia).
+    // s0 + offset:       Ia memory store operation
+    // s0 + offset + 1:   Ib memory store operation
+    // s0 + offset + 1:   Ic memory store operation
+
+    uint32_t offset = 0;
+
+    while (offset < s1) {
+        FF ib(0);
+        FF ic(0);
+        uint32_t mem_op_b(0);
+        uint32_t mem_op_c(0);
+        uint32_t mem_idx_b(0);
+        uint32_t mem_idx_c(0);
+        uint32_t rwb(0);
+        uint32_t rwc(0);
+        auto clk = mainTrace.size();
+
+        FF ia = callDataMem.at(s0 + offset);
+        uint32_t mem_op_a(1);
+        uint32_t mem_idx_a = d0 + offset;
+        uint32_t rwa = 1;
+
+        // Storing from Ia
+        ffMemory.at(mem_idx_a) = ia;
+        storeAInMemTrace(mem_idx_a, ia);
+
+        if (s1 - offset > 1) {
+            ib = callDataMem.at(s0 + offset + 1);
+            mem_op_b = 1;
+            mem_idx_b = d0 + offset + 1;
+            rwb = 1;
+
+            // Storing from Ib
+            ffMemory.at(mem_idx_b) = ib;
+            storeBInMemTrace(mem_idx_b, ib);
+        }
+
+        if (s1 - offset > 2) {
+            ic = callDataMem.at(s0 + offset + 2);
+            mem_op_c = 1;
+            mem_idx_c = d0 + offset + 2;
+            rwc = 1;
+
+            // Storing from Ic
+            ffMemory.at(mem_idx_c) = ic;
+            storeCInMemTrace(mem_idx_c, ic);
+        }
+
+        mainTrace.push_back(Row{
+            .avmMini_clk = clk,
+            .avmMini_ia = ia,
+            .avmMini_ib = ib,
+            .avmMini_ic = ic,
+            .avmMini_mem_op_a = FF(mem_op_a),
+            .avmMini_mem_op_b = FF(mem_op_b),
+            .avmMini_mem_op_c = FF(mem_op_c),
+            .avmMini_rwa = FF(rwa),
+            .avmMini_rwb = FF(rwb),
+            .avmMini_rwc = FF(rwc),
+            .avmMini_mem_idx_a = FF(mem_idx_a),
+            .avmMini_mem_idx_b = FF(mem_idx_b),
+            .avmMini_mem_idx_c = FF(mem_idx_c),
+        });
+
+        if (s1 - offset > 2) { // Guard to prevent overflow if s1 is close to uint32_t maximum value.
+            offset += 3;
+        } else {
+            offset = s1;
+        }
+    }
+}
+
+// RETURN opcode with direct memory access, i.e.,
+// return M_F[s0:s0+s1]
+// Simplified version with exclusively memory load operations into
+// intermediate registers and then values are copied to the returned vector.
+// TODO: Implement the indirect memory version (maybe not required)
+// TODO: taking care of flagging this row as the last one? Special STOP flag?
```

Comment:
yep i think this is what we will have in the longer term - when is last is set the stop flag must be active / be at that part of the bytecode